trunner/test_runner.py

```diff
@@ -191,29 +194,34 @@ def run_tests(self, tests: Sequence[TestOptions]):
             def set_reboot_flag(tests, idx, result):
                 # If the test is successful and the next test doesn't require loading via
                 # the plo we don't want to reboot the entire device (to speed up the test execution).
-                # There are three exceptions to this rule:
+                # There are five exceptions to this rule:
+                # 0. The yaml config has been configured with the reboot parameter set to True.
                 # 1. Runner runs in the "nightly" mode when we are not concerned about slow execution.
-                # 2. The test has failed.
+                # 2. The last executed test has failed
                 # 3. We have to enter the bootloader in order to load applications.
+                nonlocal reboot_next
```

Comment:
I don't like how stateless function was made stateful one with nonlocal variable - this makes code flow much less readable.

 also:
 - don't write how many exceptions (maybe "a few"), right now the number is invalid. I think if the code would be readable enough, the comment might be simplified (no need to explicitly write down each case reflected in the code)
 - (connected with the above) split the single multiline `if` with internal `or`'s into multiple ifs (one if per rule exception) to make it more readable
 - IMHO it would be more forward to call set_reboot_flag with `idx + 1` and internally use only idx (the first if can be moved to the caller) 
 - change name to `update` as it doesn't always set the `should_reboot` field

generic idea for the outer code:
```python
if result.is_fail():
    last_test_failed = True
    fail += 1
elif resut.is_success():
    last_test_failed = False
else: # result.is_skip()
    skip += 1

# [...]

if  idx + 1 != len(tests):
   update_reboot_flag(tests, idx + 1, last_test_failed)  # no need for result(?)
   # please note that this call can also be moved to the beginning of the loop (probably with if idx > 0) 
   #  - not sure if it would make it more or less readable (but more intuitive - you would be changing current test options, not the next one)
   # if the function would be short enough, consider removing it and placing code directly in the loop
```